internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,198 @@
+	if err != nil {
+		batch.Add(err)
+		return batch.Compile()
+	}
+
+	metricsUnderTest := buildMetricNames(prefix, clientOrServer)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
+		if _, ok := metricsUnderTest[name]; !ok {
+			continue
+		}
+
+		batch.Add(lintMetric(name))
+		batch.Add(validateName(name, prefix, clientOrServer))
+
+		labels := map[string]struct{}{}
+		for _, metric := range m.GetMetric() {
+			for _, label := range metric.Label {
+				labels[*label.Name] = struct{}{}
+			}
+		}
+		batch.Add(validateLabels(name, prefix, clientOrServer, labels))
+
+		// after we validate the metric, delete it from the set of all
+		// expected metrics so that we can track if any were not found.
+		delete(metricsUnderTest, name)
+	}
+
+	if len(metricsUnderTest) > 0 {
+		batch.Add(fmt.Errorf("%w: %v", errNotFound, keysFrom(metricsUnderTest)))
+	}
+
+	return batch.Compile()
+}
+
+func keysFrom(metrics map[string]struct{}) []string {
+	missing := []string{}
+	if len(metrics) > 0 {
+		for k := range metrics {
+			missing = append(missing, k)
+		}
+	}
+	return missing
+}
+
+// buildMetricNames creates a set of all the expected metrics names for the given prefix
+// that should be registered with Prometheus as defined in the baseplate spec.
+// The prefix will be either thrift, http, or grpc.
+func buildMetricNames(prefix, clientOrServer string) map[string]struct{} {
+	suffixes := []string{"latency_seconds", "requests_total", "active_requests"}
+	names := map[string]struct{}{}
+	for _, suffix := range suffixes {
+		names[prometheus.BuildFQName(prefix, clientOrServer, suffix)] = struct{}{}
+	}
+	return names
+}
+
+// validateName checks the following:
+//   1) the metric name has at least 3 parts separated by "_".
+//     The parts are <namespace>_<client/server>_<name>_<suffix>, where suffix is optional.
+//     Ref: https://prometheus.io/docs/practices/naming
+//   2) the metric name has the correct prefix.
+//   3) the metric contains either "client" or "server" as the second part.
+func validateName(name, prefix, clientOrServer string) error {
+	var batch errorsbp.Batch
+
+	const (
+		separator = "_"
+		partCount = 3
+	)
+	parts := strings.SplitN(name, separator, partCount)
+	if len(parts) < partCount {
+		batch.Add(fmt.Errorf("%w: name: %q part count: %d", errLength, name, len(parts)))
+		return batch.Compile()
+	}
+
+	if got, want := parts[0], prefix; got != want {
+		batch.Add(fmt.Errorf("%w: name: %q prefix: %q", errPrefix, name, prefix))
+	}
+
+	if got, want := parts[1], clientOrServer; got != want {
+		batch.Add(fmt.Errorf("%w: name: %q", errClientServer, name))
+	}
+	return batch.Compile()
+}
+
+func validateLabels(name, prefix, clientOrServer string, gotLabels map[string]struct{}) error {
+	var batch errorsbp.Batch
+
+	wantLabels := buildLables(name, prefix, clientOrServer)
+	if diff := cmp.Diff(gotLabels, wantLabels); diff != "" {
+		batch.Add(fmt.Errorf("%w: (-got +want)\n%s", errDiffLabels, diff))
```

Comment:
since this function can return at most one error, I think using a batch here is overkill. this line can simply be:

```go
return fmt.Errorf("%w: (-got +want)\n%s", errDiffLabels, diff)
```

Reply:
good catch. _fixing_